core/backend/src/CheckpointManager.ts

```diff
@@ -268,16 +268,59 @@ export class V1CheckpointManager {
   }
 }
 
+/** @internal */
+class RetryableError {
+  constructor(public innerErr: Error) {}
+}
+
 /** @internal  */
 export class CheckpointManager {
   public static readonly onDownloadV1 = new BeEvent<(job: DownloadJob) => void>();
   public static readonly onDownloadV2 = new BeEvent<(job: DownloadJob) => void>();
   public static getKey(checkpoint: CheckpointProps) { return `${checkpoint.iModelId}:${checkpoint.changeset.id}`; }
 
+  /**
+   * This function expects that the "functionToRetry" throws an instance of RetryableError and just any other error otherwise if its a non retryable error. This is to allow for the case where
+   * we may not want to retry every type of error produced by the function.
+   * @param numAttempts number of times to attempt the function
+   * @param functionToRetry function to call during attempts
+   */
+  private static async withAttempts(numAttempts: number, functionToRetry: () => Promise<void>): Promise<void> {
```

Comment:
I'm worried about what this will do to the progress function. Please let me look at this more closely.

Reply:
see https://github.com/iTwin/itwinjs-core/pull/3912